ow_power_system/src/PowerSystemPack.cpp

```diff
@@ -0,0 +1,375 @@
+// TODO (JW): By its nature, this file is designed to be copied. It should
+// be explicitely (un)licensed as public domain or CC0.
+#include <chrono>
+#include <ctime>
+#include <fstream>
+#include <iomanip>
+#include <map>
+#include <sstream>
+#include <vector>
+#include <numeric>
+#include <math.h>
+#include <algorithm>
+#include <ros/package.h>
+#include <std_msgs/Float64.h>
+#include <std_msgs/Int16.h>
+#include <ow_lander/lander_joints.h>
+
+#include <ros/ros.h>
+#include <sensor_msgs/JointState.h>
+#include <PrognoserFactory.h>
+
+#include "Messages/MessageBus.h"
+#include "Messages/ProgEventMessage.h"
+#include "Messages/ScalarMessage.h"
+#include "ModelBasedAsyncPrognoserBuilder.h"
+
+#include "PowerSystemPack.h"
+#include "PowerSystemNode.h"
+
+using namespace PCOE;
+
+const auto START_TIME       = MessageClock::now();
+//const int NUM_NODES         = 8; // NOTE: Should equal m_num_nodes in PowerSystemPack.h
+
+// The index use to access temperature information.
+// This might change to median SOC or RUL index or fixed percentile.
+//
+static constexpr int TEMPERATURE_INDEX = 1;
+
+double EoD_events[NUM_NODES][3];
```

Comment:
I don't like this as a global variable. There are ways to keep it internal to the classes, and that should be preferred. The first that comes to mind is to make EoD_events a data struct, for example
```C++
struct EoDEvents {
  double rul;
  double soc;
  double temperature;
}
```
Then make this struct a member of the `PowerNodeHandler` object I proposed earlier. Then pass a reference to the `PowerNodeHanlder`'s instance of `EoDEvents` into `PredictionHandler::processMessage`, so the function can directly write data to it.

Reply:
Now that I've reached the end of the non-optional comments, I'll start tackling the overhaul involving all the optional ones. I agree that the current implementation is quite flawed.

However, to my understanding I can't modify the parameters to `PredictionHandler::processMessage`; I believe the `MessageBus` that calls it is handled by GSAP and would require a modification on their end to work with it. One of the bigger hurdles I'm considering is how to get such a reference into that function since that was the primary reason I resorted to a global variable for the first implementation anyways.

Reply:
> However, to my understanding I can't modify the parameters to `PredictionHandler::processMessage`; I believe the `MessageBus` that calls it is handled by GSAP and would require a modification on their end to work with it. One of the bigger hurdles I'm considering is how to get such a reference into that function since that was the primary reason I resorted to a global variable for the first implementation anyways.

That's right, I forgot the function's prototype is API defined. But that shouldn't be a problem. You'll just have to pass a reference to `EoD_Events` into `PredictionHandler`, so it can save it as a reference member variable, which can then be modified each time `processMessage` is called.

Reply:
That's actually what I spent this morning trying to do, but you can't declare an uninitialized reference variable, so I wasn't able to save it outside of the constructor. I'll take another pass at it though; at least I know I was thinking along the right track.

Reply:
I've got this working now; the global variable is no longer needed. I can separate `PredictionHandler` into its own source file now that it no longer depends on a global var (requested in a different comment), so I'll get on that.